src/main/java/com/commercetools/sync/customers/CustomerSync.java

```diff
@@ -0,0 +1,235 @@
+     *
+     * @param customerSyncOptions the container of all the options of the sync process including the CTP project
+     *                            client and/or configuration and other sync-specific options.
+     * @param customerService  the customer service which is responsible for fetching/caching the Customers from the
+     *                         CTP project.
+     * @param typeService  the type service which is responsible for fetching/caching the Types from the CTP project.
+     * @param customerGroupService  the customer group service which is responsible for fetching/caching the
+     *                              CustomerGroups from the CTP project.
+     */
+    protected CustomerSync(@Nonnull final CustomerSyncOptions customerSyncOptions,
+                           @Nonnull final CustomerService customerService,
+                           @Nonnull final TypeService typeService,
+                           @Nonnull final CustomerGroupService customerGroupService) {
+        super(new CustomerSyncStatistics(), customerSyncOptions);
+        this.customerService = customerService;
+        this.referenceResolver = new CustomerReferenceResolver(getSyncOptions(), typeService, customerGroupService);
+        this.batchValidator = new CustomerBatchValidator(getSyncOptions(), getStatistics());
+    }
+
+    /**
+     * Iterates through the whole {@code customerDrafts} list and accumulates its valid drafts to batches.
+     * Every batch is then processed by {@link CustomerSync#processBatch(List)}.
+     *
+     * <p><strong>Inherited doc:</strong>
+     * {@inheritDoc}
+     *
+     * @param customerDrafts {@link List} of {@link CustomerDraft}'s that would be synced into CTP project.
+     * @return {@link CompletionStage} with {@link CustomerSyncStatistics} holding statistics of all sync
+     *         processes performed by this sync instance.
+     */
+    @Override
+    protected CompletionStage<CustomerSyncStatistics> process(@Nonnull final List<CustomerDraft> customerDrafts) {
+        final List<List<CustomerDraft>> batches = batchElements(customerDrafts, syncOptions.getBatchSize());
+        return syncBatches(batches, completedFuture(statistics));
+
+    }
+
+    @Override
+    protected CompletionStage<CustomerSyncStatistics> processBatch(@Nonnull final List<CustomerDraft> batch) {
+        final ImmutablePair<Set<CustomerDraft>, CustomerBatchValidator.ReferencedKeys> result =
+            batchValidator.validateAndCollectReferencedKeys(batch);
+
+        final Set<CustomerDraft> validCustomerDrafts = result.getLeft();
+        if (validCustomerDrafts.isEmpty()) {
+            statistics.incrementProcessed(batch.size());
+            return completedFuture(statistics);
+        }
+
+        return referenceResolver
+            .populateKeyToIdCachesForReferencedKeys(result.getRight())
+            .handle(ImmutablePair::new)
+            .thenCompose(cachingResponse -> {
```

Comment:
Looks like you are right because it's not needed in here, most structure of this code coming from the existing code base, for other implementations maybe we could check and refactor all similar unnecessary mappings in a different PR.

Reply:
it is not easy to return stages from the handle, if I change it, I need to change in many places, this practice does not need an update for now.